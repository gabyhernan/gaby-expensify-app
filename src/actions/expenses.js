// action generators for expenses
import uuid from 'uuid';
import database from '../firebase/firebase';


// Steps of action generators
// 1. Component calls action generator
// 2. Action generator returns an object
// 3.  Component takes objct & passes it to dispatch
// 4. Redux store runs the reducers & it changes

// We need to tweak how our action generators work to integrate with db
// With asynchronous actions things look a little bit different
// 1. Component calls action generator
// 2. Action generator returns a FUNCTION
// 3. Component dispatches FUNCTION
// 4. Funcion runs (has the ability to dispatch other actions
// & do whatever it wants, so we put DB code here, and we have ability
// to dispatch another action, a standard one that returns an object ,
// and that will manipulate the redux store)

//  By default Redux DOES NOT allow you to dispatch functions.
// This is why you have to set up a module which is a piece of redux middleware
// that adds support for this behavior

// ADD_EXPENSE /
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
   expense
});

// change defaults to our db call
export const startAddExpense = ( expenseData = {}) => {
  // NOTE: we could set up defaults in function arg but we are
  // gonna do it in a different way so u know that it can be done like this also

  // we ar going to return a function
  return (dispatch) => {  // only works cuz the middleware has been setup
    // gets called internally by redux
    const {
      description = '',
      note = '',
      amount = 0, // destructuring from expenseData
      createdAt = 0
    } = expenseData

    const expense = { description, note, amount, createdAt };
    // 1. writing some data to firebase
    // returning this so we can start a promise chain
  return database.ref('expenses').push(expense).then( (ref) =>{
 // we have to dispatch action from up above otherwise store WONT CHANGE eva
    dispatch(addExpense({
      id: ref.key,
      ...expense
    }));
    }) // Note that you have to dispatch this guy now instead of addExpense
    // wherever u were using it in ur app

  };
};


// REMOVE_EXPENSE
export const removeExpense = ( {id} = {} ) => ({
    type: 'REMOVE_EXPENSE',
      id
});


// START REMOVE EXPENSE
export const startRemoveExpense = ( {id } = {} ) => {
  return (dispatch) => {
      return database.ref(`expenses/${id}`).remove().then( (ref) =>{
          dispatch(removeExpense({ id }));
      }).catch( (e) => console.log('Error fetching data', e));
  }
}


// EDIT_EXPENSE
export const editExpense = ( id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = ( id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then( () => {
      dispatch(editExpense(id, updates));

    }).catch( (e) => console.log('Error fetching data', e));
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// START_SET_EXPENSES - asynchronous action that will fetch data
// 1. fetch all expense data once
// 2. parse that data into an array
// 3. Dispatch SET_EXPENSES
// 4.
export const startSetExpenses = () => {
  return (dispatch) => {

   return database.ref('expenses').once('value')
  .then( (snapshot) => {
   const expenses = [];
    // iterate over all of the child snapshots & toss them in this array
    snapshot.forEach( (childSnapshot) => {
      // every snapshot has access to key which is that uniqueId firebase creates
      expenses.push({
        id: childSnapshot.key,
        // spreads out what comes back from child snapshot
        // so we dont have to write it every single time
        ...childSnapshot.val()
      });

      dispatch(setExpenses(expenses));
    })
  })
  .catch( (e) => {
    console.log('error fetching data', e);
  });

  };
};
