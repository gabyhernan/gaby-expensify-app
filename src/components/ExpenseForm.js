import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


// now the big picture goal here is to use local component state
// to track the changes to all of these inputs

// only when the user ACTUALLY SUBMITS the form will we do something with the info
// when the submit the form we send it off to rudux store to either edit or create
// a new expense

// we use moment.js library to work with date & time
const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
  // so the goal is to  take these four state values & ONLY use the defaults
  // seen here if no EXPENSE was passed down, thar's going to make sure the add
  // expense page still works , but if an expense was PASSED DOWN we want
  // to start these off at those values in the EditExpensePage
  // to get that done we have to look to the props, so we have to define our
  // state in the constructor function to access the props

  constructor(props) {
    super(props);

    this.state = {
   description: props.expense ? props.expense.description : '' ,
 // makes sure that page works if we are editing an expense or if we are adding
    note: props.expense ? props.expense.note : '',
    amount: props.expense ? (props.expense.amount / 100).toString() : '',
// it's a number we want a string and we want it in cents so we divide by 100
// then we convert it to a string
    createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
   // we call moment since if it was created we want the time and date
 // when it was created now the current time
    calendarFocused: false,// setting it as not focused initially
    error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState( () => ({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    // set number validation for form
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { // .match takes our regex expression in
   // !amount allows user to clear the value if they write the wrong amount initally
     this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) { // putting it in if statement to prevents user for being
      // able to clear date value which is bad UX experience
    this.setState( () => ({ createdAt }));
    }
  }

  onFocusChange = ( {focused }) => {
    this.setState( () => ({ calendarFocused: focused }));
  }


  onSubmit = (e) => { // for our forms we want to access e
    // to preventDefault and that full page refresh
    e.preventDefault();

  if(!this.state.description || !this.state.amount) {
   // set error state equal to 'please provide description & amount'
   this.setState( () => ({ error: 'Please provide description & amount' }));
  } else {
    // clear the error
    this.setState ( () => ({ error: '' }));
    console.log('submitted'); // set up action when the form is submitted
    this.props.onSubmit({
      description: this.state.description,
      amount: parseFloat(this.state.amount, 10) * 100 ,// since we are working with cents
      //setting it in right format parseFloat keeps ddecimals in place
      createdAt: this.state.createdAt.valueOf(), // currently a moments object
      // valueOf is a moments method that changes it to unix timestamp which JS works
      // in milliseconds
      note: this.state.note
    });
  }

  }

  render () {
   return (
    <div>
 { this.state.error && <p> {this.state.error} </p> }

      <form onSubmit={this.onSubmit}>
        <input
          className="text-input"
          type="text"
          placeholder="description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
        // have to set  our own validation for numbers
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
         />

        <SingleDatePicker
          date={this.state.createdAt} // moment object represents where you want to start
          onDateChange={this.onDateChange} // function we create with moment instance when someone picks
          // new date from calendar
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1} // shows 1 month at a time
          isOutsideRange={ () => false } // if we had an extensive lodging app we
          // could set a function specify the days but since we just want to be
          // able to go back to the days we set it as false
           />


         <textarea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
         >
          </textarea>

          <button > Add Expense </button>
      </form>
    </div>
    );
  }
}
