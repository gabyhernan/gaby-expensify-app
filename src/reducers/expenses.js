

// Expenses Reducer

const expensesReducerDefaultState = [];

export const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
     return  [ // ...state refers to all of the existing items in array
     ...state ,
     action.expense // adding on this new item to a NEW array
     ];
     case 'REMOVE_EXPENSE':
     return state.filter(( {id}) => id !== action.id);
  // if they are not equal it will result in true meaning id will be kept
// if they are equal , will result in false & item will be discarded
    case 'EDIT_EXPENSE':  // want go through every expense in array
    // when we find match we want to edit it
    return state.map((expense) => {
      if (expense.id === action.id){
        return {
          // grab all properties of existing one using spread operator
          ...expense,
          ...action.updates // overwrites existing properties
          // with the ones user passes in
        }
      }
       else {
        return expense;
      }
    });

     case 'SET_EXPENSES':
    return action.expenses;


    default:
      return state;
  }
};
