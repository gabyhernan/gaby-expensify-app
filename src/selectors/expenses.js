// selectors - aka the one we show to the screen
import moment from 'moment';

// timestamps - dates
// are any positive or negative integer value ex. 204, -30, 2000

// like our year system the'yre counting based off some metric
// timestamps are counting in milliseconds

// what does time stamp 0 represent
// It represents January 1st, 1970 midnight (known as unix epic or epoch)
// this specific date is the starting point for all of our timestamps

// so positive numbers are gonna come afterwards
// negative numbers are going to be before

// we are going to be using timestamps as a way to store t
//imezone independent time data


// GET VISIBLE EXPENSES
// we are just going to pass the data into a single function
// it is going to filter & sort the data returning the visible expenses
// & thats what we will be able to show to the screen

// we have to destructure the filters creating vars for the invidual ones
 const getVisibleExpenses = (expenses, {text = '', sortBy, startDate,endDate}) => {
  // use .filter to return a subset of all the expenses
  return expenses.filter( (expense) => { //
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
// check if user provided startDate, return true if there isnt since we dont want
// filtering to occur
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true ;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    //  includes method checks a given match inside of strings
    // convert both strings to lower case to create a case insensitive search

// we use sort on arrays
// ok to use sort directly when it comes to sorting simple strings
// but for arrays of more complex data like objext there is no inherent sorting value
// we are going to need to define whats known as a compare function
// this lets us write a lil code to determine which item should come first
// we return -1 if a shold come first
// we return 1 if b should come first
// sort like filter gets called on the array and it returns the array

    // if all of these are true, we have a complete match & we want to return
    // true from filter
    return startDateMatch && endDateMatch && textMatch
    // if all these are true then the item will be kept in the array
  }).sort((a, b) => {
    if (sortBy === 'date') {
      // if a.createdAt is less than b.created we can go ahead and toss b first
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      // put the ones with a greater amount first
      return a.amount >= b.amount ? -1 : 1;
    }
  })
};


export default getVisibleExpenses;
