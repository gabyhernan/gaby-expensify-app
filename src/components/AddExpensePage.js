import React from 'react';
import {connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {   // props.dispatch(addExpense(expense));
      this.props.addExpense(expense);
 // on ReactRouter we have history methods, inside we .push is how we can
// programatically change pages , takes single string args which is ur path
      this.props.history.push('/');
    }

  render() {
  return (
  <div>
    <ExpenseForm
    onSubmit={this.onSubmit} /> {/* the add expense page needs to be able to actually dispatch the
    given action to the redux store  */}

    This is from my add expense page component
  </div>
  );
  }


}




// very similar to mapStateToProps
// but instead of working with state, it works with dispatch
const mapDispatchToProps = (dispatch) => ({
  // goal is to return an object
 // define props that are going to call dispatch
    addExpense: (expense) => dispatch(addExpense(expense))
 // now we have the same functionality, but we have a compoment that
  // is now more testable
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
// first arg is map state to props , pass as undefined since we dont need for this one

// second function is called mapDispatchToProps -
// it is a way to return your dispatcher functions allowing you to abstract
// them away from the component itself
