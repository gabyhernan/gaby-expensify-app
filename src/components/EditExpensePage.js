import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// Refactor to be a class based comp
// Setup mapDispatchToProps - editExpense, removeExpense


export class EditExpensePage extends React.Component {

  onSubmit = (expense) => {
// Dispatch the action to edit the expense,Then redirect to dashboard page
      this.props.startEditExpense(this.props.expense.id, expense);
// console.log('updated', expense);
 // on ReactRouter we have history methods, inside we .push is how we can
// programatically change pages , takes single string args which is ur path
      this.props.history.push('/');
          }

  onClickRemoveExpense = () => {
     this.props.startRemoveExpense({ id: this.props.expense.id});
     this.props.history.push('/');
    }

  render () {
      return (
  <div>
        <ExpenseForm
   // we expect that the previous data we created in AddExpense page to be here
// to get that done we have to take the expense and pass it down
          expense={this.props.expense}
          onSubmit={this.onSubmit}
          />

    <button onClick={this.onClickRemoveExpense} > Remove </button>


  </div>
  );

  }
};

const mapDispatchToProps = (dispatch, props) => {
  return {
  startEditExpense: (id, expense) =>  dispatch(startEditExpense(id, expense)) ,
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
  }

};
// this time we want to set up a mapStateToProps
// we want to be able to give the compo the current expense object
const mapStateToProps = (state, props) => {
  // we are searching the expenses array for an expense whose ID matches
  // this props.match.params.id

  // we have access to the props right here as the second argument
  // we can take some of the current props that were passed into the HOC comp
  // & we can use them to calculate the props that we want to add on
  return {
    expense: state.expenses.find( (expense) => expense.id === props.match.params.id)
 // only return the object IF their ids match
  };
};



export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
