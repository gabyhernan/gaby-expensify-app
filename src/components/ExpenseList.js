import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import {getVisibleExpenses} from '../selectors/expenses'

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
          <p> No expenses </p>
        ) : (
          props.expenses.map( (expense) => {
      return <ExpenseListItem key={expense.id} {...expense}/>
    })
        )
    }
    <h1> Expense List </h1>



   </div>
  );

const mapStateToProps = (state) => {
// maps the component state  to the component props
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
    // use the selector so our data is actually changing
  };
};

// creating a new const for the higher order component
// not a commont pattern to name your higher order comp
export default connect(mapStateToProps)(ExpenseList);
// the first argment is where provide the information about what we want to connect
// the component doesn't need everything from the store usually needs a subset

// so we define a function that lets us determine what info from the store
// we want our component to have
// store's state gets passed in as first arg
// we then return an object , in which we can put any key value pairs


// IMP TO KNOW
// When you connect a component to the redux store it's reactive which means
// that as the store changes your component is going to get re-rendered with those
// new values

// This is great cause it allows to create very simple compoonents
// the component doesn't need to worry about store.subscribe or store.getState
// it doesn't have to manage any component state to manage that data ,
// it's all done for us by react redux

