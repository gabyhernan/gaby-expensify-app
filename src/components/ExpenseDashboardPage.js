import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';


const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
    This is from my dashboard component
  </div>
  );



  export default ExpenseDashboardPage;
