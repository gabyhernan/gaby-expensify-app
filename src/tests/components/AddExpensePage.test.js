import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage }from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

// if you find yourself with a lot of duplicate code on your test cases
// Jest provides lifecycle methods like afterAll, afterEach.
// beforeAll, beforeEach

// using these lifecycle methods we are to able to set up the
// spies & component, then each test case can worry about
// just using those

let startAddExpense, history, wrapper;
// going to be using beforeEach to run some code before each test case
beforeEach(() => {
   startAddExpense = jest.fn();
   history = { push: jest.fn() };
   wrapper = shallow(<AddExpensePage startAddExpense={startAddExpense} history={history} />);
}); // we can now use spies & wrappers throughtout all test cases
// without writing them every single time

test('should render AddExpensePage correctly', () => {
  // set up some spies for props

  expect(wrapper).toMatchSnapshot();

});

// test that when the form gets submitted both of our spies get passed with correct info
test('should handle addExpense onSubmit', () => {


  // call the function that gets passed into expense form
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

  // make some assertions
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1]);
});
