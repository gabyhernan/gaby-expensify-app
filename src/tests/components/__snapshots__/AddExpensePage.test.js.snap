// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
   
  This is from my add expense page component
</div>
`;
